x/mongo/driver/topology/pool.go

```diff
@@ -735,38 +751,59 @@ func (p *pool) removeConnection(conn *connection, reason reason, err error) erro
 	return nil
 }
 
-// checkIn returns an idle connection to the pool. If the connection is perished or the pool is
-// closed, it is removed from the connection pool and closed.
+// checkIn returns an idle connection to the pool. It calls checkInWithCallback internally.
 func (p *pool) checkIn(conn *connection) error {
-	if conn == nil {
-		return nil
-	}
-	if conn.pool != p {
-		return ErrWrongPool
-	}
+	return p.checkInWithCallback(conn, func() (reason, bool) {
+		if mustLogPoolMessage(p) {
+			keysAndValues := logger.KeyValues{
+				logger.KeyDriverConnectionID, conn.driverConnectionID,
+			}
 
-	if mustLogPoolMessage(p) {
-		keysAndValues := logger.KeyValues{
-			logger.KeyDriverConnectionID, conn.driverConnectionID,
+			logPoolMessage(p, logger.ConnectionCheckedIn, keysAndValues...)
 		}
 
-		logPoolMessage(p, logger.ConnectionCheckedIn, keysAndValues...)
-	}
-
-	if p.monitor != nil {
-		p.monitor.Event(&event.PoolEvent{
-			Type:         event.ConnectionCheckedIn,
-			ConnectionID: conn.driverConnectionID,
-			Address:      conn.addr.String(),
-		})
-	}
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionCheckedIn,
+				ConnectionID: conn.driverConnectionID,
+				Address:      conn.addr.String(),
+			})
+		}
 
-	return p.checkInNoEvent(conn)
+		r, perished := connectionPerished(conn)
+		if !perished && conn.pool.getState() == poolClosed {
+			perished = true
+			r = reason{
+				loggerConn: logger.ReasonConnClosedPoolClosed,
+				event:      event.ReasonPoolClosed,
+			}
+		}
+		return r, perished
```

Comment:
I'm concerned about the amount of duplicate code the `checkInWithCallback` function creates. There are currently multiple sources of truth for check-in logging, events, and perished connection behavior, depending on which function is called. We should find a way to remove the duplicated code.